catalyst-gateway/bin/src/settings/str_env_var.rs

```diff
@@ -0,0 +1,272 @@
+//! Processing for String Environment Variables
+use std::{
+    env::{self, VarError},
+    fmt::{self, Display},
+    str::FromStr,
+};
+
+use strum::VariantNames;
+use tracing::{error, info};
+
+/// An environment variable read as a string.
+#[derive(Clone)]
+pub(crate) struct StringEnvVar {
+    /// Value of the env var.
+    value: String,
+    /// Whether the env var is displayed redacted or not.
+    redacted: bool,
+}
+
+/// Ergonomic way of specifying if a env var needs to be redacted or not.
+pub(super) enum StringEnvVarParams {
+    /// The env var is plain and should not be redacted.
+    Plain(String, Option<String>),
+    /// The env var is redacted and should be redacted.
+    Redacted(String, Option<String>),
+}
+
+impl From<&str> for StringEnvVarParams {
+    fn from(s: &str) -> Self {
+        StringEnvVarParams::Plain(String::from(s), None)
+    }
+}
+
+impl From<String> for StringEnvVarParams {
+    fn from(s: String) -> Self {
+        StringEnvVarParams::Plain(s, None)
+    }
+}
+
+impl From<(&str, bool)> for StringEnvVarParams {
+    fn from((s, r): (&str, bool)) -> Self {
+        if r {
+            StringEnvVarParams::Redacted(String::from(s), None)
+        } else {
+            StringEnvVarParams::Plain(String::from(s), None)
+        }
+    }
+}
+
+impl From<(&str, bool, &str)> for StringEnvVarParams {
+    fn from((s, r, c): (&str, bool, &str)) -> Self {
+        if r {
+            StringEnvVarParams::Redacted(String::from(s), Some(String::from(c)))
+        } else {
+            StringEnvVarParams::Plain(String::from(s), Some(String::from(c)))
+        }
+    }
+}
+
+/// An environment variable read as a string.
+impl StringEnvVar {
+    /// Read the env var from the environment.
+    ///
+    /// If not defined, read from a .env file.
+    /// If still not defined, use the default.
+    ///
+    /// # Arguments
+    ///
+    /// * `var_name`: &str - the name of the env var
+    /// * `default_value`: &str - the default value
+    ///
+    /// # Returns
+    ///
+    /// * Self - the value
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// #use cat_data_service::settings::StringEnvVar;
+    ///
+    /// let var = StringEnvVar::new("MY_VAR", "default");
+    /// assert_eq!(var.as_str(), "default");
+    /// ```
+    pub(super) fn new(var_name: &str, param: StringEnvVarParams) -> Self {
+        let (default_value, redacted, choices) = match param {
+            StringEnvVarParams::Plain(s, c) => (s, false, c),
+            StringEnvVarParams::Redacted(s, c) => (s, true, c),
+        };
+
+        match env::var(var_name) {
```

Comment:
What do you think to rewrite this `match` and nested `if, else` statements into something like:
```Rust
match (env::var(var_name), redacted, choices) {
 (Ok(value), true, Some(choices)) => {...}
 (Ok(value), false, Some(choices)) => {...}
}
```
Yes it will increase the number of `match` arms but probably will reduce the complexity and readability of each arm

Reply:
Could be cleaner, will look at it. :+1:

Reply:
The new version is as clean as I can make it, and the code not be very redundantly defined.
The conditionals are now only around the logging, and that's because of limitations with the info! macros themselves not allowing the use of dynamic fields.